utilities/ldap_sync.sql

```diff
@@ -25,118 +152,763 @@ Changes in this version:
+    except Exception as e:
+        raise Exception("Unable to find/parse exa.get_connection(ctx.LDAP_CONNECTION).password", exa.meta.script_name, str(e), abortMessage)
+        
+    try:
+        encoding = "utf8"  #may depend on ldap server, try latin1 or cp1252 if you get problems with special characters
+    except Exception as e:
+        raise Exception("Unable to set encoding = utf8", exa.meta.script_name, str(e), abortMessage)
+    
+    #----------------------------------------
+    # Sets a network timeout of 15 seconds to connect to LDAP
+    #----------------------------------------
+    try:
+        ldap.set_option(ldap.OPT_NETWORK_TIMEOUT, 30.0)
+    except Exception as e:
+	    raise Exception("Failure on ldap.set_option(ldap.OPT_NETWORK_TIMEOUT, 5.0)", exa.meta.script_name, str(e), abortMessage)
 	
-		# Emits the results of the specified attributes
-		for result in results:
-			result_dn = result[0]
-			result_attrs = result[1]
-
-	  		if ctx.ATTR in result_attrs:
-				for v in result_attrs[ctx.ATTR]:
-			             ctx.emit(ctx.SEARCH_STRING, ctx.ATTR, str(v).decode(encoding))#add .decode if there are special characters
-
-	except ldap.LDAPError, e:
-		if e.message['desc'] == 'No such object':
-		      ctx.emit(ctx.SEARCH_STRING, ctx.ATTR, 'No such object')
-		else:
-		      raise ldap.LDAPError(e.message['desc'])
-		
-	finally:
-		ldapClient.unbind_s()		
+	#----------------------------------------
+	# The below line is only needed when connecting via ldaps
+	#----------------------------------------
+    try:
+        ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)   # required options for SSL without cert checking
+    except Exception as e:
+        raise Exception("Failed! ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER", exa.meta.script_name,  str(e), abortMessage)
+        
+    #========================================
+    # Begin LDAP interaction
+    #========================================  
+    try:
+        ldapClient = ldap.initialize(uri)   # Connects to LDAP
+    except Exception as e:
+        raise Exception(f"Ldap initialization failed. Check the uri from uri = {uri}" , exa.meta.script_name, str(e), abortMessage)
+    
 
-/
+        
+#######################################
+# MAIN LOGIC
+#######################################
+
+    #----------------------------------------
+    # Authenticates with connection credentials
+    #----------------------------------------
+    try:
+        ldapClient.bind_s(user, password)
+    except Exception as e:
+	        raise Exception(f"Failed: ldapClient.bind_s(user, password) for user {user}" , exa.meta.script_name, str(e), abortMessage)
+	    
+    #---------------------------------------
+    # Python 3 can not handle bytes, so we decode to chg bytes to string
+    #---------------------------------------  
+    try:
+        global results
```

Comment:
you can pull the results in front of the try, then you don't need to declare it global